group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/ModernGroup.java

```diff
@@ -388,10 +386,25 @@ public Map<String, TopicMetadata> computeSubscriptionMetadata(
         subscribedTopicNames.forEach((topicName, count) -> {
             TopicImage topicImage = topicsImage.getTopic(topicName);
             if (topicImage != null) {
+                Map<Integer, Set<String>> partitionRacks = new HashMap<>();
+                topicImage.partitions().forEach((partition, partitionRegistration) -> {
+                    Set<String> racks = new HashSet<>();
+                    for (int replica : partitionRegistration.replicas) {
+                        Optional<String> rackOptional = clusterImage.broker(replica).rack();
```

Comment:
This seems more like an optimization rather than a bug fix, correct? There isn't any data loss even if we don't trigger a rebalance. Also, this rebalance event isn't mentioned in [KIP-848](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=217387038#KIP848:TheNextGenerationoftheConsumerRebalanceProtocol-GroupEpoch-Triggerarebalance).

Reply:
On another note, moving a partition can trigger a rack change, correct? If so, could this cause issues for users who have deployed an auto-rebalance tool? It might lead to a series of rebalances.

Reply:
@chia7712 It is actually a gap in KIP-848 to include https://cwiki.apache.org/confluence/display/KAFKA/KIP-881%3A+Rack-aware+Partition+Assignment+for+Kafka+Consumers.

> On another note, moving a partition can trigger a rack change, correct? If so, could this cause issues for users who have deployed an auto-rebalance tool? It might lead to a series of rebalances.

It should not cause issues because the assignment should only change if the consumers are not aligned with racks anymore. Otherwise, it should stay the same.

Reply:
@dajac thanks for your response. 

> It should not cause issues because the assignment should only change if the consumers are not aligned with racks anymore. Otherwise, it should stay the same.

Yes, that makes sense. However, how do we determine if assignment changes require a rebalance or not? For instance, if partition_a currently has replicas in [rack_0, rack_1, rack_2] and there are no consumers in rack_2, moving the replica from rack_2 to rack_1 should NOT trigger a rebalance, correct? It's not an ideal use case, just something for discussion :smiley:

Reply:
In this case, it would trigger a "rebalance" by bumping the group epoch and it would trigger the computation of a new assignment for the group. In this particular example, the assignment should not change because, as you said, there is no reason to change it. Hence it would effectively be a no-op for the consumers.